### ISSUE TITLE
More modular codebase

### ISSUE THREAD

**Issue by vemv** [clojure-lsp/clojure-lsp#658]
Hi there,

In [this](https://clojurians.slack.com/archives/CPABC1H61/p1636412265027200) thread we noted that the project's dep tree is quite dense. That surely makes sense on its own, but when using clojure-lsp as a dependency in another tooling project (let's say hypothetically, cider-nrepl, refactor-nrepl, etc), then it's bit of a hairy thing to place in the dep tree.

(Especially as [clojure-emacs](https://github.com/clojure-emacs) projects typically go through mranderson - it's better to not pass a lot of stuff to it.)

If wanting to consume clojure-lsp as a library, probably I would be most interested on its "functional core", namely pure functions that would take clj-kondo analysis as an input, and perform e.g. rewrite-clj operations as an output.

LSP interaction should not take place at all when performing these - nREPL is its own protocol, so having more moving pieces seems a bit redundant/risky.

Likewise I'd be able to pass clj-kondo analyses (as in-mem data or custom directories) as an argument. i.e. building a lsp/kondo cache would be ideally decoupled/optional. This way I can pass guaranteedly up-to-date analyses that don't require a manual CLI setup.

(I've followed this model in a project of mine, worked very well)

So my suggestion would boil down to:

* Split the project in sub-modules of sorts, such that one can depend on a custom subset
* Have the core functionality mostly decoupled from IO-impure edges, most notably LSP
* Be able to pass custom clj-kondo data as inputs to the core functionality

And the possible outcome would be:

* The project can back or complement the implementation of more tooling projects.

From my side I don't have an immediate plan to go ahead and change the clojure-emacs projects I contribute to, but certainly this would seem a great thing to have in the horizon. Perhaps 1y from now I could go ahead, if gradually clojure-lsp becomes more and more fit for such an integration, would be awesome of course!

-V

**Comment by ericdallo** [comment]
Thanks for the issue, it makes sense Indeed.
I made some refactor some months ago to have the db passed from outside to inside, so this can help with a refactor or a component replacement maybe.

**Comment by ericdallo** [comment]
@vemv good news, I started working last week on some huge refactor that got merged today, we now have 2 submodules the lib and the cli, it's a start for a more modular codebase, this one made possible remove all the CLI stuff including lsp4j, and other deps that doesn't make sense for the API!

**Comment by vemv** [comment]
Very nice, thanks so much!

Exciting news.

`lib`'s dep tree looks far more digestible now. Looks like there's still some road ahead, with Timbre, clj-kondo and datalevin.

From my side there's no particular rush anyway.

**Comment by ericdallo** [comment]
Thank you! 
Yes, I intend to break those even more, also, I'm considering replacing datalevin with transit, just like kondo, we are having some graalvm issues and some inconsistency from time to time unfortunately

**Comment by Cyrik** [comment]
I'd be a big fan of switching to transit. I'd love to consume LSP's kondo cache, but the datalevin dependency is too large.
So if I can help in any way with the switch let me know!

**Comment by ericdallo** [comment]
Thanks @Cyrik ! I intend to work on this soon, not next release, but next, I'll basically do some performance tests and consistency tests of the cache, but it should be no problem for transit I guess

**Comment by ericdallo** [comment]
@Cyrik you may be interested on [this](https://github.com/clojure-lsp/clojure-lsp/issues/703)

**Comment by Cyrik** [comment]
I'll have a look, thanks for pinging me!

**Comment by ericdallo** [comment]
@vemv After #1117 we feel confortable enough to close this as we removed lots of deps using the clojure impl of lsp4clj, most concerns of the original issue's description are fixed/gone, but please feel free to open a new issue for/if any still present concerns or suggestions, thank you for all the help!

c/c @mainej

**Comment by vemv** [comment]
Thanks much! Looking forward to do something with it.

Might create a (simpler) follow-up issue
